### PR TITLE
Aws ec2 inventory fix

### PR DESCRIPTION
**Pull request by willthames**
##### SUMMARY
Improve AWS EC2 inventory plugin

* Better grouping by tags
* Allow hostname tag to be set without providing a value
* Fix EXAMPLE documentation formatting

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
aws_ec2 inventory plugin 

##### ANSIBLE VERSION
```
ansible 2.6.0 (devel 0214a85382) last updated 2018/03/21 15:11:32 (GMT +1000)
  config file = /home/will/tmp/ansible/aws_ec2_inventory/ansible.cfg
  configured module search path = [u'/home/will/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/will/src/ansible/lib/ansible
  executable location = /home/will/src/ansible/bin/ansible
  python version = 2.7.14 (default, Feb 27 2018, 20:43:24) [GCC 7.3.1 20180130 (Red Hat 7.3.1-2)]
```

### CODE REVIEW

**Comment by willthames** [comment]
@bcoca - I can see no test cases for `keyed_groups` but this will affect *all* inventory plugins that rely on keyed_groups, so I would appreciate your review.

File: lib/ansible/plugins/inventory/__init__.py
Comment (s-hertel) [inline lib/ansible/plugins/inventory/__init__.py]: I think this is still necessary. I'm running into `Invalid group name
format, expected string or list of strings, got: <type 'unicode'>` and
`Invalid group
name format, expected string or list of strings, got: <class 'ansible.parsing.yaml.objects.AnsibleUnicode'>` testing different keyed groups. Putting this back fixes it.

File: lib/ansible/plugins/inventory/__init__.py
Comment (s-hertel) [inline lib/ansible/plugins/inventory/__init__.py]: This no longer works consistently with values. It works with values that are in {key: value} format, but not [{key: value}]. If a user wants to create specific groups from ProductCodes/BlockDeviceMappings/ElasticGpuAssociations/NetworkInterfaces/SecurityGroups/Tags (i.e. anything that is a list of dicts), the value is ignored and all hosts with the key are added to the group.
reproducer:
```
plugin: aws_ec2
regions:
  - us-east-1
keyed_groups:
  - key: 'SecurityGroups'
    value: {'GroupId': 'sg-xxxxxxxx'}
```
returns all instances in us-east-1 in the SecurityGroups group, regardless of the group id.

**Review by s-hertel — COMMENTED** [review]
Looks good from an aws_ec2 perspective. This allows to make groups like 
`tag_key_value`
whereas before the format was
`tag___key____value__` and make all tags into tag_key_value groups rather than requiring a value for each.

File: lib/ansible/plugins/inventory/__init__.py
Comment (bcoca) [inline lib/ansible/plugins/inventory/__init__.py]: we shoudl not rely on convert_bare as we plan on removing it

File: lib/ansible/plugins/inventory/__init__.py
Comment (bcoca) [inline lib/ansible/plugins/inventory/__init__.py]: This does not allow for prefixes w/o _, if the user wants to add/ignore _ he should be able to

File: lib/ansible/plugins/inventory/__init__.py
Comment (bcoca) [inline lib/ansible/plugins/inventory/__init__.py]: removing this will create a bunch of uncaught exceptions down the line

File: lib/ansible/plugins/inventory/aws_ec2.py
Comment (bcoca) [inline lib/ansible/plugins/inventory/aws_ec2.py]: redundant, this info is already in the description

**Review by bcoca — CHANGES_REQUESTED** [review]
~~-1 from the base perspective, the changes reduce user options and create pitfalls for bad values~~
rethought this a bit and came with an alternative that provides the functionality sought

File: lib/ansible/plugins/inventory/__init__.py
Comment (willthames) [reply lib/ansible/plugins/inventory/__init__.py]: Is there an alternative - t.template converts the result back to the correct data type, whereas t.do_template does not, and I don't really want to have to reimplement the conversion logic.

File: lib/ansible/plugins/inventory/__init__.py
Comment (willthames) [reply lib/ansible/plugins/inventory/__init__.py]: In the previous code, if you didn't specify a prefix your group names would just start with an underscore, so I fail to see how this is worse.

File: lib/ansible/plugins/inventory/__init__.py
Comment (willthames) [reply lib/ansible/plugins/inventory/__init__.py]: I can wrap the call to `to_safe_group_name` in the same exception handling.

File: lib/ansible/plugins/inventory/aws_ec2.py
Comment (willthames) [reply lib/ansible/plugins/inventory/aws_ec2.py]: Fair, but notes are more obvious when reading the docs. I can remove this (I hadn't noticed it in the description when I added it)

File: lib/ansible/plugins/inventory/__init__.py
Comment (willthames) [reply lib/ansible/plugins/inventory/__init__.py]: Did this work before this change?

File: lib/ansible/plugins/inventory/__init__.py
Comment (willthames) [reply lib/ansible/plugins/inventory/__init__.py]: Yes, I'm not sure I meant to remove this completely.

File: lib/ansible/plugins/inventory/__init__.py
Comment (willthames) [reply lib/ansible/plugins/inventory/__init__.py]: Can I use `t.template('%s%s%s' % (t.environment.variable_start_string, template, t.environment.variable_end_string), disable_lookups=True)` instead then?

File: lib/ansible/plugins/inventory/__init__.py
Comment (willthames) [reply lib/ansible/plugins/inventory/__init__.py]:
```
keyed_groups:
  - key: 'security_groups|json_query("[].group_id")'
    prefix: 'security_groups'
```

works as expected. I'm not really sure how `value` actually provides filtering here, or why you'd want that vs. using `filters` section.

File: lib/ansible/plugins/inventory/__init__.py
Comment (willthames) [reply lib/ansible/plugins/inventory/__init__.py]: What exceptions are generated by `to_safe_group_name` - it just calls `re.sub` doesn't it? (assuming I don't cause the errors @s-hertel mentions through lack of ensuring that groups is a list rather than a stringtype)

Comment (bcoca) [reply lib/ansible/plugins/inventory/__init__.py]: the exceptions will happen in the plays when people try to access invalid group names

File: lib/ansible/plugins/inventory/__init__.py
Comment (bcoca) [reply lib/ansible/plugins/inventory/__init__.py]: nvmd, missed the composition below

Comment (bcoca) [reply lib/ansible/plugins/inventory/__init__.py]: but you actually remove teh prefix?

File: lib/ansible/plugins/inventory/__init__.py
Comment (s-hertel) [reply lib/ansible/plugins/inventory/__init__.py]: ~This is broken on devel,~ you can disregard the comment. Not related to your PR.
I'm not sure if and how this worked. LGTM.

File: lib/ansible/plugins/inventory/__init__.py
Comment (willthames) [reply lib/ansible/plugins/inventory/__init__.py]: There shouldn't be any unsafe group names - the conversion to safe group names just happens slightly later.

**Comment by bcoca** [comment]
see #38126

**Comment by s-hertel** [comment]
I rebased to drop the commit for lib/ansible/plugins/inventory/__init__.py since #38126 was merged instead. This fixes aws_ec2 hostnames using tags and should be backported to 2.5. I have tested using the branch https://github.com/s-hertel/ansible/commits/aws_ec2_inventory_tests_using_pr/38126_and_pr/37689. Once this is merged I will update my tests PR.

@bcoca can you update your review on this, please?

**Review by bcoca — APPROVED** [review]

**Comment by flowerysong** [comment]
Hmm. I prefer my version (#35880); while I might be missing some context that makes prefixing the hostname with the tag name and an underscore desirable, I feel like most people won't want that to happen.

**Comment by bcoca** [comment]
@flowerysong these are not either/or, inventory plugins can allow you to define inventory_hostname differently, it doesn't make sense for all of them, but in this case it might. 

I'll leave that up to others in the community that currently use aws, unlike myself.

**Comment by willthames** [comment]
@flowerysong I do prefer yours - I'll merge yours and then rebase this on top.

**Comment by s-hertel** [comment]
rebuild_merge

**Pull request merged**
